src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,767 @@
+                    int lastLevel = 2 * (numLeaves - levelCount);
+                    assert lastLevel >= 0;
+                    if (lastLevel != 0)
+                    {
+                        // Last level is partially filled, so we must rotate the leaf FPs to match.  We do this here, after loading
+                        // at read-time, so that we can still delta code them on disk at write:
+                        long[] newLeafBlockFPs = new long[numLeaves];
+                        System.arraycopy(leafBlockFPs, lastLevel, newLeafBlockFPs, 0, leafBlockFPs.length - lastLevel);
+                        System.arraycopy(leafBlockFPs, 0, newLeafBlockFPs, leafBlockFPs.length - lastLevel, lastLevel);
+                        leafBlockFPs = newLeafBlockFPs;
+                    }
+                    break;
+                }
+
+                levelCount *= 2;
+            }
+        }
+
+        // Reused while packing the index
+        try (RAMIndexOutput writeBuffer = new RAMIndexOutput("PackedIndex"))
+        {
+            // This is the "file" we append the byte[] to:
+            List<byte[]> blocks = new ArrayList<>();
+            byte[] lastSplitValue = new byte[bytesPerValue];
+            int totalSize = recursePackIndex(writeBuffer, leafBlockFPs, splitValues, 0, blocks, 1, lastSplitValue, false);
+            // Compact the byte[] blocks into single byte index:
+            byte[] index = new byte[totalSize];
+            int upto = 0;
+            for (byte[] block : blocks)
+            {
+                System.arraycopy(block, 0, index, upto, block.length);
+                upto += block.length;
+            }
+            assert upto == totalSize;
+
+            return index;
+        }
+    }
+
+    /**
+     * lastSplitValue is the split value previously seen; we use this to prefix-code the split byte[] on each
+     * inner node
+     */
+    private int recursePackIndex(RAMIndexOutput writeBuffer, long[] leafBlockFPs, byte[] splitValues, long minBlockFP, List<byte[]> blocks,
+                                 int nodeID, byte[] lastSplitValue, boolean isLeft) throws IOException
+    {
+        if (nodeID >= leafBlockFPs.length)
+        {
+            int leafID = nodeID - leafBlockFPs.length;
+
+            // In the unbalanced case it's possible the left most node only has one child:
+            if (leafID < leafBlockFPs.length)
+            {
+                long delta = leafBlockFPs[leafID] - minBlockFP;
+                if (isLeft)
+                {
+                    assert delta == 0;
+                    return 0;
+                }
+                else
+                {
+                    assert nodeID == 1 || delta > 0 : "nodeID=" + nodeID;
+                    writeBuffer.writeVLong(delta);
+                    return appendBlock(writeBuffer, blocks);
+                }
+            }
+            else
+            {
+                throw new IllegalStateException();
+//                return 0;
```

Comment:
TODO: cleanup, etc.

Reply:
Note: In all the testing I have done on this, I haven't managed to create an unbalanced tree that gets here. As a result, I have added the `IllegalStateException` (with a message now). I will continue to investigate this to see if it's possible and we will get some performance testing done of this to see if we get here on larger, real world, datasets.